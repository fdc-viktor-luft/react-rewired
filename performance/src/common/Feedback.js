// @flow

import React, { Component } from 'react';

type FeedbackStatus = 'good' | 'neutral' | 'bad';

const _get = (): { [string]: FeedbackStatus } => {
    try {
        return JSON.parse((localStorage.getItem('feedback'): any)) || {};
    } catch {
        return {};
    }
};

const _set = (id: string, status: FeedbackStatus) => {
    const current = _get();
    current[id] = status;
    localStorage.setItem('feedback', JSON.stringify(current));
};

const FeedbackStorage = {
    get: _get,
    set: _set,
};

type FeedbackProps = {| id: string |};
type FeedbackState = { status?: FeedbackStatus };

type FeedbackStatusProps = {| onClick?: void => void |};
const Good = ({ onClick }: FeedbackStatusProps) => (
    <div className="status" style={{ color: '#66ff6d' }}>
        <span onClick={onClick} aria-label="good" role="img" aria-labelledby="me">
            &#x1F603;
        </span>
    </div>
);
const Neutral = ({ onClick }: FeedbackStatusProps) => (
    <div className="status">
        <span onClick={onClick} aria-label="neutral" role="img" aria-labelledby="me">
            &#x1F610;
        </span>
    </div>
);
const Bad = ({ onClick }: FeedbackStatusProps) => (
    <div className="status" style={{ color: '#ff4f56' }}>
        <span onClick={onClick} aria-label="bad" role="img" aria-labelledby="me">
            &#x1F61E;
        </span>
    </div>
);

export class Feedback extends Component<FeedbackProps, FeedbackState> {
    state: FeedbackState = { status: FeedbackStorage.get()[this.props.id] };

    choose = (status: FeedbackStatus) => () => FeedbackStorage.set(this.props.id, status) || this.setState({ status });

    renderStatus = (status: FeedbackStatus) => {
        if (status === 'good') return <Good />;
        if (status === 'neutral') return <Neutral />;
        if (status === 'bad') return <Bad />;
    };

    render = () => {
        const { status } = this.state;

        return (
            <div className="feedback">
                {status ? (
                    this.renderStatus(status)
                ) : (
                    <>
                        <Good onClick={this.choose('good')} />
                        <Neutral onClick={this.choose('neutral')} />
                        <Bad onClick={this.choose('bad')} />
                    </>
                )}
            </div>
        );
    };
}
