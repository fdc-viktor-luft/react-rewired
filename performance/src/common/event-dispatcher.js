// @flow

const nativeInputValueSetter = (Object.getOwnPropertyDescriptor(window.HTMLInputElement.prototype, 'value'): any).set;

const change = (inputElement: any, value: string) => {
    nativeInputValueSetter.call(inputElement, value);
    const e = new Event('input', { bubbles: true });
    inputElement.dispatchEvent(e);
};

const changeById = (id: string, value: string) => {
    const input = document.getElementById(id);
    if (input && input.tagName === 'input') change(input, value);
};

export const EventDispatcher = { change, changeById };
