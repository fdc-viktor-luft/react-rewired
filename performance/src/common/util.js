// @flow

const randomColor = () => '#xxxxxx'.replace(/x/g, () => Math.floor(Math.random() * 16).toString(16));

export const Util = { randomColor };
