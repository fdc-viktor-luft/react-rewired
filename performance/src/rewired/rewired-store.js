// @flow

import { Wired, type WiredNode } from 'react-rewired';

type OneInputState = { value: string, color: string };
const initialOneInput = {
    value: '',
    color: '#eeff8e',
};

const deeplyNested = Wired.node({
    a: Wired.node({
        b: Wired.node({ c: Wired.node({ d: Wired.node({ e: Wired.node({ f: Wired.node({ g: 0 }) }) }) }) }),
    }),
});

const initialManyFlat: { [string]: number } = 'abcdefghijklmnopqrstuvwxyz'
    .split('')
    .reduce((p, c) => (p[c] = 0) || p, {});

type LifeCodingState = {};

type State = {
    oneInput: WiredNode<OneInputState>,
    deeplyNested: typeof deeplyNested,
    manyFlat: typeof initialManyFlat,
    liveCoding: WiredNode<LifeCodingState>,
    manyComponents: number,
};

export const RewiredStore = Wired.store<State>({
    oneInput: Wired.node(initialOneInput),
    deeplyNested,
    manyFlat: Wired.node(initialManyFlat),
    liveCoding: Wired.node({}),
    manyComponents: 0,
});

window.RewiredStore = RewiredStore;
