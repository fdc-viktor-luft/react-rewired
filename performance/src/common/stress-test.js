// @flow

import React from 'react';
import { EventDispatcher } from './event-dispatcher';

// optimized as much as possible
// call this function without arguments to generate a new uuid
type UuidParam = void;
const uuid = (a: UuidParam): string =>
    a ? (a ^ ((Math.random() * 16) >> (a / 4))).toString(16) : (([1e7]: any) + -1e3 + -4e3).replace(/[018]/g, uuid);

const doXTimes = (times: number, handler: number => void) => () => {
    let counter = 0;
    const start = performance.now();
    const _handler = () => {
        counter++;
        handler(counter);
        if (counter === times) console.log('duration ms', Math.ceil(performance.now() - start));
    };
    for (let i = 0; i < times; i++) window.setTimeout(_handler, 0);
};

const changeXTimesInput = (times: number, inputId: string) => () => {
    const input = document.getElementById(inputId);
    if (input && input.tagName.toLowerCase() === 'input')
        doXTimes(times, index => EventDispatcher.change(input, String(index) + ' - ' + uuid()))();
};

// 100'000 changes take for both ~ 200 sec
// 10'000 changes take for both ~ 15 - 16 sec --> should be the ideal choice for demonstration
// 1'000 changes take for both ~ 1.5 - 1.6 sec
const ButtonChangeX = ({ inputId, times }: {| inputId: string, times: number |}) => (
    <button onClick={changeXTimesInput(times, inputId)}>Change {times} times</button>
);

const ButtonDoX = ({ onClick, times }: {| onClick: number => void, times: number |}) => (
    <button onClick={doXTimes(times, onClick)}>Do {times} times</button>
);

export const StressTest = { ButtonChangeX, ButtonDoX };
