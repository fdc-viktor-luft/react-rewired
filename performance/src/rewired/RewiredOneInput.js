// @flow

import React from 'react';
import { RewiredStore } from './rewired-store';
import { RewiredDisplayValues } from './RewiredDisplayValues';
import { EventUtil } from '../common/event-util';
import { StressTest } from '../common/stress-test';
import { Util } from '../common/util';

const onChangeInput = EventUtil.inputHandler(v =>
    RewiredStore.set({ oneInput: { value: v, color: Util.randomColor() } })
);

type RewiredOneInputStoreProps = {| value: string, color: string |};
type RewiredOneInputOwnProps = {||};
export type RewiredOneInputProps = {| ...RewiredOneInputStoreProps, ...RewiredOneInputOwnProps |};

export const RewiredOneInputContainer = ({ value, color }: RewiredOneInputProps) => (
    <div className="one-input">
        <RewiredDisplayValues />
        <input
            id="rewired-input"
            type="text"
            value={value}
            style={{ backgroundColor: color }}
            onChange={onChangeInput}
        />
        <StressTest.ButtonChangeX inputId="rewired-input" times={5000} />
    </div>
);

export const RewiredOneInput = RewiredStore.wire<RewiredOneInputStoreProps, RewiredOneInputOwnProps>(
    RewiredOneInputContainer,
    ({ oneInput }) => ({
        value: oneInput.value,
        color: oneInput.color,
    })
);
