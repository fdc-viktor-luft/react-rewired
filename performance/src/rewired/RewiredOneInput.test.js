// @flow

import React from 'react';
import { Spy } from 'spy4js';
import { shallow } from 'enzyme';
import { RewiredOneInput, RewiredOneInputContainer, type RewiredOneInputProps } from './RewiredOneInput';
import { RewiredStore } from './rewired-store';
import { Util } from '../common/util';

describe('RewiredOneInput', () => {
    it('maps properties to container', () => {
        // initialize the store with any properties
        RewiredStore.set({ oneInput: { value: 'testValue', color: '#test' } });
        // now shallow mount the wired component
        expect(shallow(<RewiredOneInput />)).toMatchInlineSnapshot(`
<RewiredOneInputContainer
  color="#test"
  value="testValue"
/>
`);
    });
});

describe('RewiredOneInputContainer', () => {
    const Mock$RewiredStore = Spy.mock(RewiredStore, 'set');
    const Mock$Util = Spy.mock(Util, 'randomColor');

    const getWrapper = (overrides?: $Shape<{ ...RewiredOneInputProps }>) =>
        shallow(<RewiredOneInputContainer value="test-value" color="#abcdef" {...overrides} />);

    it('applies value and color to the output', () => {
        expect(getWrapper()).toMatchSnapshot();
    });

    it('handles input change', () => {
        // prepare mocks
        const preventDefault = new Spy();
        Mock$Util.randomColor.returns('#fff000');

        // mount the wrapper
        const wrapper = getWrapper();

        // change the input
        wrapper.find('input').simulate('change', { target: { value: 'foo' }, preventDefault });

        // assertions
        preventDefault.wasCalled(1);
        Mock$RewiredStore.set.hasCallHistory({ oneInput: { value: 'foo', color: '#fff000' } });
    });
});
