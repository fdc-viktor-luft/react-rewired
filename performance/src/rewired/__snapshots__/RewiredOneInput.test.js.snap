// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`RewiredOneInputContainer applies value and color to the output 1`] = `
<div
  className="one-input"
>
  <Wired(DisplayValues) />
  <input
    id="rewired-input"
    onChange={[Function]}
    style={
      Object {
        "backgroundColor": "#abcdef",
      }
    }
    type="text"
    value="test-value"
  />
  <ButtonChangeX
    inputId="rewired-input"
    times={10000}
  />
</div>
`;
